[Wip] Improve Threadpool QUWI throughput

benaadams opened the pull request:
11% Improvement for regular queuing (1.6sec over 10M QUWI)
19.4% improvement for high thread count queuing (MinWorkerThreads=500)  (7.2sec over 10M QUWI)

Test project: https://gist.github.com/benaadams/b022934e62a3ac1c4f261be3216b1111

10M threadpool queues and executes. Changed items in red, `ExecutionContext.Run` highlighted and list past `ExecutionContext.Restore` for relative comparison.

Threadpool QUWI before
![threadpool-before](https://aoa.blob.core.windows.net/aspnet/threadpool-before.png)

Threadpool QUWI after 2nd Update
![threadpool-after-2](https://aoa.blob.core.windows.net/aspnet/threadpool-after-2.png)

benaadams commented:
Verifying

stephentoub commented on src/mscorlib/src/System/Threading/ThreadPool.cs:
```diff
@@ -129,10 +129,34 @@ internal void Remove(T e)
         {
             private const int INITIAL_SIZE = 32;
-            internal volatile IThreadPoolWorkItem[] m_array = new IThreadPoolWorkItem[INITIAL_SIZE];
+            internal volatile PaddedWorkItem[] m_array = new PaddedWorkItem[INITIAL_SIZE];
```
Why are you 64-byte padding the items in the queue?  The queue is owned by a single thread, and all other threads need to take a lock to access it.  The owning thread also needs to take a lock when there are fewer than two items in the queue.  The only case where you'd have contention this would help with is if there was a thread stealing concurrently with the owning thread pushing/popping on a list with at least two elements.  At that point, they're already some distance apart, though not necessarily a full cache line.  Have you shown that this change makes a notable improvement?  It does so at the expense of effectively increasing the size of every work item by 56 bytes on 64-bit, since every work item reference to be stored now consumes 64 bytes instead of 8 bytes (plus the size of the work item object itself).

stephentoub commented on src/mscorlib/src/System/Threading/ThreadPool.cs:
```diff
@@ -129,10 +129,34 @@ internal void Remove(T e)
+            public volatile QueueSegment Segment;
+        }
+
+        // Enusre full cache line occupied so no false sharing with other indicies
```
Enusre => Ensure in all of these cases

Technically this isn't ensuring it, either, as there are systems with cache line sizes greater than 64 bytes.  But they're less common.

stephentoub commented on src/mscorlib/src/System/Threading/ThreadPool.cs:
```diff
@@ -414,23 +438,23 @@ private bool TrySteal(out IThreadPoolWorkItem obj, ref bool missedSteal, int mil
         {
             // Holds a segment of the queue.  Enqueues/Dequeues start at element 0, and work their way up.
-            internal readonly IThreadPoolWorkItem[] nodes;
+            internal readonly PaddedWorkItem[] nodes;
```
Same question here: why are you padding this?  What contention are you trying to prevent? Two threads operating on the same segment are already going to be contending on indexes, so it's not clear to me how padding each node element and the indexes and the Next field really helps.  You have numbers to show that it does?  This is again adding significant overhead in memory.  Previously on 64-bit each of these nodes arrays would have been ~2K (8 bytes * 256 elements). Now they're each ~16K (64 bytes * 256 elements).

ericeil replied:
Additionally, now when accessing these we need to touch one cache line per workitem, vs. one line per 8 work items in the current code.  I'd expect this to result in quite a lot more cache churn, and thus much slower execution, in many scenarios.

benaadams commented:
@stephentoub as you pointed out don't think padding the items is helpful

Still working on it - hot spots are Dequeue and TrySteal

![](https://aoa.blob.core.windows.net/aspnet/threadpool.png)

benaadams commented:
Looking closer the main effect may just be looping the queues (many threads) with mainly empty queues.

omariom commented:
@benaadams I think you should start from finding  false sharings.[ In this article](http://www.codeproject.com/Articles/51553/Concurrency-Hazards-False-Sharing) shown how to use VS for that. Not sure if it possible on Windows 10,but on Win 7 it was.

benaadams commented:
@omariom there is false sharing in stealing; however the current implementation even with the false sharing is pretty hard to improve on. Still iterating, though looking at something quite different than the PR in this current state.

omariom commented:
Interesting optimization would be to find all the places (on hot pathes) where volatile reads/writes are unnecessary, replace them with plain reads and use `Volatile`  class in the other places. 
It may help on ARM where volatile reads/writes are implemented as  fairly expensive full barriers.

benaadams commented:
> Interesting optimization would be to find all the places (on hot pathes) where volatile reads/writes are unnecessary, replace them with plain reads and use Volatile class in the rest. 

There are some areas where there are multiple of the same types in a row e.g. 3x volatile write or 3x volatile read which may cause over use of memory barriers. Will try it and measure the impact.

Although such a change does make me a little uncomfortable :worried:

benaadams commented:
Bit better, getting a 6% improvement for QueueUserWorkItem throughput for 10M work items (4 core 1 socket)
With 9% improvement for `set COMPlus_ThreadPool_ForceMinWorkerThreads=500`

Still investigating.

benaadams commented:
Now 14% improvement for `set COMPlus_ThreadPool_ForceMinWorkerThreads=500` (31.2s vs 36.6s)

benaadams commented:
10M threadpool queues and executes. Changed items in red, `ExecutionContext.Run` highlighted and list to `System.Random.Sample` for relative comparison.

Threadpool QUWI before
![threadpool-before](https://aoa.blob.core.windows.net/aspnet/threadpool-before.png)

Threadpool QUWI after
![threadpool-after](https://aoa.blob.core.windows.net/aspnet/threadpool-after.png)

benaadams commented:
Test code https://gist.github.com/benaadams/b022934e62a3ac1c4f261be3216b1111

It also allocates 2112 bytes per 255 queued items in discarded `QueueSegment`s (85MB per 10M) - which give a process equilibrium at 300MB-400MB memory use vs 50MB equilibrium without these allocations.

Caching a QueueSegment as it gets dropped off the tail and the reusing it as it for a new head avoids these allocations but its also not entirely straightforward with the concurrency flows; so not perusing that at this stage.

I believe the changes in this PR should not alter any of the concurrency behaviour.

prajwal-aithal commented:
@dotnet-bot test Linux ARM Emulator Cross Debug Build

benaadams commented:
@dotnet-bot test Linux ARM Emulator Cross Release Build

benaadams commented:
@dotnet-bot test

benaadams commented:
@dotnet-bot test this please

benaadams commented:
Added QueueSegment reuse as commit, needs tests re rerunning

omariom commented:
@benaadams 

What happens if a consumer thread is preempted [at this point](https://github.com/benaadams/coreclr/blob/40d1b2de8578230585521f092714131d76004b0e/src/mscorlib/src/System/Threading/ThreadPool.cs#L794) right after it's made a local copy of `queueTail` segment, then other threads  consume up  the segment => it is assigned to `hazardSegment`, and  then the next segment is consumed up =>  `hazardSegment` becomes `freeSegment`? Now the segment the first thread has on its stack is going to be used when it  and  it is free at the same time. 
Will the queue survive such scenario?

benaadams commented:
@omariom could you update your link to last commit?

omariom commented:
@benaadams  Replaced it with description.

benaadams commented:
@omariom removed the Segment reuse, its a tricky area and should probably be examined in isolation

benaadams commented:
@dotnet-bot test Linux ARM Emulator Cross Debug Build
@dotnet-bot test Linux ARM Emulator Cross Release Build

benaadams commented:
@dotnet-bot test Linux ARM Emulator Cross Debug Build

benaadams commented:
Threadpool QUWI after 2nd Update
![threadpool-after-2](https://aoa.blob.core.windows.net/aspnet/threadpool-after-2.png)

11% Improvement for regular queuing (1.6sec over 10M QUWI)
19.4% improvement for high thread count queuing (MinWorkerThreads=500)  (7.2sec over 10M QUWI)

benaadams commented:
High threadcount time is all spent in Native code; so can't address that more in the managed stack

![high-threadcount](https://aoa.blob.core.windows.net/aspnet/threadpool-after-highthread.png)

benaadams commented:
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test

stephentoub commented:
One of the difficulties with performance work on the thread pool is that there are many situations, and improvements to one benchmark can hurt others.  I'm not saying that's the case with your changes (I've not reviewed them), but I worry when I see a single benchmark being used as the sole guard rail for making these kinds of changes.  For example, ThreadPool.QueueUserWorkItem always queues to the global queue, so it looks like your benchmark will always have empty local queues, yet changes in this PR are affecting how threads search for work in the local queues, e.g. whether to start from a deterministic or random position.

benaadams commented:
Native costs high in ThreadpoolMgr::WorkerThreadStart
4 Cores
![](https://aoa.blob.core.windows.net/aspnet/threadpoolnative.png)

8 HT Cores (4 Cores)
![](https://aoa.blob.core.windows.net/aspnet/threadpoolnative-3.png)

4 Core summary
![](https://aoa.blob.core.windows.net/aspnet/threadpoolnative-2.png)

benaadams commented:
Low thread count costs

![](https://aoa.blob.core.windows.net/aspnet/threadpoolnative-4.png)

benaadams commented:
@stephentoub working on some tests https://github.com/benaadams/ThreadPoolTaskTesting

Need to add `.Continuation`s and `Task<T>` produces output as the following; tests seem ok?

```
Testing 2,621,440 calls, with GCs after 262,144 calls.
Operations per second on 4 Cores
                                                                       Parallelism
                            Serial          2x         16x         64x        512x
QUWI                      11.207 M    11.133 M    10.367 M    11.995 M    12.728 M
- Depth    2               9.840 M     8.643 M     9.060 M     9.680 M     9.849 M
- Depth   16              10.316 M     9.967 M    10.142 M    10.262 M    10.167 M
- Depth   64              10.180 M    10.281 M    10.434 M    10.516 M    10.432 M
- Depth  512              10.165 M    10.404 M    10.478 M    10.561 M    10.370 M

Yielding Await           777.115 k     1.777 M     4.330 M     4.297 M     4.442 M
- Depth    2               1.119 M     2.632 M     5.864 M     5.989 M     5.865 M
- Depth   16               3.170 M     7.062 M     8.017 M     8.157 M     8.037 M
- Depth   64               3.801 M     7.728 M     8.371 M     8.307 M     8.406 M
- Depth  512               4.159 M     7.878 M     7.588 M     8.163 M     7.624 M

Async Awaited            612.523 k   938.077 k     3.019 M     3.187 M     3.215 M
- Depth    2             967.841 k     1.327 M     4.680 M     4.848 M     4.772 M
- Depth   16               1.952 M     3.633 M     7.471 M     8.064 M     7.857 M
- Depth   64               2.066 M     4.224 M     8.416 M     8.179 M     8.440 M
- Depth  512               2.176 M     4.377 M     6.715 M     7.126 M     8.114 M

Async PassThrough        608.733 k   945.321 k     2.988 M     3.444 M     3.413 M
- Depth    2               1.185 M     1.515 M     6.717 M     6.939 M     6.949 M
- Depth   16              10.090 M    15.450 M    54.967 M    44.132 M    51.816 M
- Depth   64              38.301 M    18.645 M   165.344 M   202.727 M   168.579 M
- Depth  512             274.373 M   524.928 M     1.028 B     1.040 B   571.730 M

Completed Awaited         23.897 M    46.998 M    88.617 M    89.584 M    84.225 M
- Depth    2              25.996 M    52.894 M    99.585 M    98.614 M    92.834 M
- Depth   16              20.297 M    40.554 M    75.178 M    75.558 M    71.595 M
- Depth   64              19.189 M    38.543 M    70.893 M    70.111 M    67.075 M
- Depth  512              18.494 M    37.213 M    69.032 M    66.975 M    67.496 M

CachedTask Awaited        23.465 M    48.319 M    87.310 M    87.725 M    87.067 M
- Depth    2              26.575 M    52.979 M    96.709 M    95.250 M    93.352 M
- Depth   16              20.586 M    40.269 M    73.974 M    75.871 M    73.815 M
- Depth   64              18.858 M    38.509 M    69.453 M    70.659 M    69.170 M
- Depth  512              18.732 M    37.364 M    69.064 M    67.129 M    67.454 M

CachedTask CheckAwait    113.367 M   229.085 M   429.308 M   424.339 M   360.767 M
- Depth    2             137.955 M   270.930 M   476.011 M   488.828 M   406.236 M
- Depth   16             152.997 M   304.613 M   574.248 M   548.361 M   427.990 M
- Depth   64              93.764 M   178.281 M   326.802 M   327.406 M   275.967 M
- Depth  512              82.238 M   163.306 M   301.808 M   293.775 M   240.028 M

CachedTask PassThrough   120.011 M   236.834 M   423.852 M   428.908 M   351.923 M
- Depth    2             199.613 M   391.400 M   700.452 M   703.723 M   518.974 M
- Depth   16             575.774 M     1.086 B     1.932 B     1.835 B     1.042 B
- Depth   64             935.627 M     1.747 B     2.879 B     2.381 B     1.185 B
- Depth  512               1.412 B     2.612 B     4.118 B     3.045 B     1.363 B
```

benaadams commented:
Deltas
 
Type                                | Serial |       2x |    16x |     64x | 512x
-------------------------- | -------| -------| -------| -------| -------|
QUWI No Queues	|	18.0%	|	12.1%	|	12.6%	|	16.4%	|	16.9%
- Depth 2	|	14.7%	|	16.0%	|	13.3%	|	13.0%	|	11.8%
- Depth 16	|	20.9%	|	17.6%	|	15.3%	|	17.9%	|	16.2%
- Depth 64	|	17.1%	|	17.5%	|	17.9%	|	15.4%	|	19.5%
- Depth 512	|	21.7%	|	16.0%	|	16.6%	|	19.3%	|	20.3%
SubTasks	|	16.6%	|	37.2%	|	6.9%	|	12.2%	|	10.4%
- Depth 2	|	38.5%	|	47.9%	|	14.2%	|	4.2%	|	6.4%
- Depth 16	|	204.6%	|	90.4%	|	130.9%	|	44.2%	|	17.2%
- Depth 64	|	139.7%	|	49.6%	|	54.8%	|	46.1%	|	19.5%
- Depth 512	|	208.4%	|	65.9%	|	26.9%	|	11.5%	|	12.0%
SubTasks Awaited	|	2.6%	|	56.9%	|	2.2%	|	13.6%	|	11.6%
- Depth 2	|	27.6%	|	93.3%	|	19.2%	|	13.7%	|	8.6%
- Depth 16	|	100.2%	|	117.3%	|	23.3%	|	15.5%	|	12.3%
- Depth 64	|	68.2%	|	89.7%	|	22.0%	|	11.7%	|	11.2%
- Depth 512	|	66.1%	|	124.6%	|	19.0%	|	19.8%	|	12.3%
QUWI Local Queues	|	19.4%	|	13.5%	|	16.9%	|	12.1%	|	12.9%
- Depth 2	|	11.8%	|	10.8%	|	11.6%	|	13.4%	|	10.8%
- Depth 16	|	13.8%	|	15.0%	|	17.1%	|	17.1%	|	15.7%
- Depth 64	|	15.9%	|	19.0%	|	17.8%	|	17.9%	|	18.0%
- Depth 512	|	17.2%	|	18.2%	|	18.4%	|	18.0%	|	18.0%
Yielding Await	|	13.3%	|	1.0%	|	12.9%	|	12.8%	|	9.3%
- Depth 2	|	7.4%	|	12.7%	|	10.2%	|	14.8%	|	9.2%
- Depth 16	|	7.7%	|	10.7%	|	-0.9%	|	8.1%	|	5.7%
- Depth 64	|	6.8%	|	16.3%	|	11.8%	|	2.7%	|	1.9%
- Depth 512	|	0.9%	|	-6.4%	|	89.6%	|	-0.3%	|	3.8%
Async Awaited	|	9.6%	|	47.4%	|	2.2%	|	2.2%	|	-1.5%
- Depth 2	|	1.7%	|	19.5%	|	2.5%	|	7.4%	|	2.9%
- Depth 16	|	11.1%	|	10.4%	|	2.9%	|	-0.5%	|	4.3%
- Depth 64	|	1.5%	|	1.0%	|	2.3%	|	3.1%	|	4.4%
- Depth 512	|	2.5%	|	-8.8%	|	10.0%	|	10.1%	|	1.3%
Async PassThrough	|	19.2%	|	43.2%	|	4.1%	|	10.1%	|	2.4%
- Depth 2	|	13.1%	|	40.9%	|	8.4%	|	4.7%	|	4.8%
- Depth 16	|	58.8%	|	51.7%	|	10.2%	|	3.4%	|	4.8%
- Depth 64	|	-3.5%	|	2.0%	|	3.5%	|	4.5%	|	9.4%
- Depth 512	|	5.6%	|	176.6%	|	4.6%	|	1.5%	|	4.0%
Completed Awaited	|	-15.2%	|	-13.9%	|	-12.2%	|	-16.1%	|	-11.8%
- Depth 2	|	-16.6%	|	-15.5%	|	-19.4%	|	-17.9%	|	-16.7%
- Depth 16	|	-10.8%	|	-12.0%	|	-12.4%	|	-11.7%	|	-15.3%
- Depth 64	|	-12.1%	|	-10.4%	|	-11.5%	|	-12.3%	|	-13.8%
- Depth 512	|	-11.5%	|	-11.0%	|	-12.2%	|	-9.2%	|	-10.6%
CachedTask Awaited	|	-15.9%	|	-17.1%	|	-17.3%	|	-15.4%	|	-16.3%
- Depth 2	|	-16.4%	|	-17.1%	|	-18.4%	|	-18.7%	|	-17.1%
- Depth 16	|	-13.2%	|	-13.2%	|	-10.9%	|	-14.2%	|	-10.5%
- Depth 64	|	-12.1%	|	-13.0%	|	-13.8%	|	-13.6%	|	-11.6%
- Depth 512	|	-10.1%	|	-11.4%	|	-8.0%	|	-9.0%	|	-9.8%
CachedTask CheckAwait	|	-6.5%	|	-6.4%	|	-3.8%	|	-4.8%	|	-3.2%
- Depth 2	|	-3.7%	|	-5.2%	|	-13.4%	|	-6.6%	|	-4.0%
- Depth 16	|	-3.4%	|	-3.4%	|	8.8%	|	-10.9%	|	-8.9%
- Depth 64	|	-1.4%	|	0.1%	|	-3.7%	|	-2.6%	|	7.6%
- Depth 512	|	0.1%	|	-0.4%	|	9.3%	|	-3.5%	|	1.9%
CachedTask PassThrough	|	1.5%	|	0.3%	|	12.4%	|	-7.5%	|	2.4%
- Depth 2	|	1.6%	|	0.8%	|	3.9%	|	4.2%	|	3.1%
- Depth 16	|	1.5%	|	-0.5%	|	1.8%	|	0.3%	|	3.6%
- Depth 64	|	0.4%	|	-3.6%	|	-0.8%	|	4.7%	|	4.5%
- Depth 512	|	1.1%	|	-0.3%	|	8.3%	|	3.9%	|	7.0%

Drops on items that don't use the pools, awaited cache/completed tasks - will look into it

stephentoub commented:
> Tests seem ok?

I did a quick skim for what you're covering; seems like a reasonable set.  Other things you might want to cover:
- I may have missed it, but I didn't see any workloads that fan out to the whole pool, e.g. like your subtasks test, except with each task creating multiple subtasks and whenall'ing on all of them.  The value there is you end up hopefully saturating the pool.
- Are there any tests that really stress stealing? e.g. a test that queues a work item which then queues another work item and blocks until it's completed.
- It looks like all of the tests have basically zero work in the body of each work item.  While that's reasonable for examining overheads, it also represents only a subset of possible workloads, and different timings can impact various aspects of the pool and its overheads.  For example, if a task queues another task, does some work, and then waits for that sub task, if the work performed was close to zero, there's a good chance this thread will be the one to process the task it queued, but if it performed non-zero work, there's a good chance another thread in the pool will end up stealing it, and that impacts overall time spent on overheads.
- Number of cores employed can have a really big impact on overheads / throughput / etc.
- As can which flavor of the GC is used.

benaadams commented:
> I may have missed it, but I didn't see any workloads that fan out to the whole pool

The Serial (1), 2x, 16x, 64x, 512x columns are essentially doing that so its done for all tests.

> test that queues a work item which then queues another work item and blocks until it's completed.

The Depth (1), 2, 16, 64, 512 rows are essentially doing that so its done for all tests.

However, I have now added some explicit fanout tests where the depth number is a whenall degree, rather than a call chain and renamed tests to be more specific whether its: chain awaited, return (passthrough) or fanout (whenall) - will work on it some more.

The total number of repetitions is `= total / (depth * parallelism)` so each entry should be the same number of calls just exercised through different dimensions. Have also added disclaimer that its for before/after testing rather than algo choice comparison.

> if the work performed was close to zero, there's a good chance this thread will be the one to process the task it queued

The steal functions do get called a lot; however the task submission the threads are balanced; will look to add unbalanced submissions, where the threads submit N to (N + M) subtasks.

> Number of cores employed can have a really big impact on overheads / throughput / etc.

I've tried it on a faster CPU that's 4 core but w/ 8 HT and have a pretty massive deterioration for some of the tests (down to > 1/5 of the slower non-HT cpu) for pre and post; which is interesting. I like a metric driven smoking gun :) Investigating...

> As can which flavor of the GC is used.

Pre and post should both release references at the same time; however I'm trying to avoid GC as much as possible as it introduces unpredictable variance in run times and would need very long runs to amortise its effects. I will look at runs with no manual GC in different flavours to ensure memory usage doesn't do anything strange. Maybe I should have an include the manual GC call time in the run time?

benaadams commented:
Probably need some AsyncLocal tests too

benaadams commented:
Documented extreme drop offs in https://github.com/dotnet/coreclr/issues/6476#issuecomment-235911516

benaadams commented:
Rebased and updated https://github.com/dotnet/coreclr/pull/6516 part

danmosemsft commented:
@benaadams what remains here to call this PR good to go?

benaadams commented:
@danmosemsft it needs to be freshened and rebased. I'll open as another PR with new results as there is a lot of noise now in this one.
